src/generate/src.rs

```diff
@@ -0,0 +1,329 @@
+}
+
+// Pretty-printing support.
+
+#[derive(Clone, Debug)]
+enum Elem {
+    Char(char, Spacing),
+    Ident(Ident),
+    Literal(Literal),
+}
+
+impl fmt::Display for Elem {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Elem::Char(c, _) => c.fmt(f),
+            Elem::Ident(i) => i.fmt(f),
+            Elem::Literal(l) => l.fmt(f),
+        }
+    }
+}
+
+impl Elem {
+    fn try_from_tt(tt: TokenTree) -> Result<Elem, (Delimiter, TokenStream)> {
+        Ok(match tt {
+            TokenTree::Group(g) => return Err((g.delimiter(), g.stream())),
+            TokenTree::Ident(i) => Elem::Ident(i),
+            TokenTree::Literal(l) => Elem::Literal(l),
+            TokenTree::Punct(p) => Elem::Char(p.as_char(), p.spacing()),
+        })
+    }
+}
+
+#[derive(Default)]
+struct Line {
+    indent: usize,
+    elems: Vec<Elem>,
+}
+
+impl fmt::Display for Line {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for _ in 0..self.indent {
+            write!(f, "    ")?;
+        }
+        for elem in &self.elems {
+            elem.fmt(f)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+pub struct Fragment {
+    before: Vec<Elem>,
+    lines: Vec<Line>,
+}
+
+impl fmt::Display for Fragment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for elem in &self.before {
+            elem.fmt(f)?;
+        }
+        for line in &self.lines {
+            write!(f, "\n{}", line)?;
+        }
+        Ok(())
+    }
+}
+
+impl Fragment {
+    fn last(&self) -> &[Elem] {
+        self.lines.last().map_or(&self.before, |line| &line.elems)
+    }
+
+    fn last_mut(&mut self) -> &mut Vec<Elem> {
+        self.lines
+            .last_mut()
+            .map_or(&mut self.before, |line| &mut line.elems)
+    }
+
+    fn push(&mut self, elem: Elem) {
+        self.last_mut().push(elem);
+    }
+
+    fn append(&mut self, other: Fragment) {
+        self.last_mut().extend(other.before);
+        self.lines.extend(other.lines);
+    }
+}
+
+impl Src {
+    pub fn pretty_print(&self) -> Fragment {
+        self.pretty_print_in_group(Delimiter::None)
+    }
+
+    fn pretty_print_in_group(&self, outer_delim: Delimiter) -> Fragment {
+        let mut frag = Fragment::default();
+        let mut prev = None;
+        for elem in self.tokens.clone().into_iter().map(Elem::try_from_tt) {
+            if let Some(prev) = prev.replace(elem.clone()) {
+                let join_symmetrical = |a: &_, b: &_| match (a, b) {
+                    // (Ok(Elem::Char2(':', ':')), Ok(Elem::Ident(_)))
+                    (Ok(Elem::Char('.', _)), _) => true,
+
+                    // FIXME(eddyb) this only makes sense in type syntax.
+                    (Ok(Elem::Char('<', _)), _) => true,
+
+                    _ => false,
+                };
+                let join = match (&prev, &elem) {
+                    (Ok(Elem::Literal(_)), Ok(Elem::Char('.', _))) => false,
```

Comment:
Comments as to what each arm tries to achieve would be neat.

Reply:
I'm fine with this whole thing being opaque, so I'm moving it to a static method on `Elem` and sticking a doc comment on it. These are all ad-hoc rules that could all be missing and everything would still work :P. (they exist only for debugging, even when rustfmt isn't in use/fails)

Reply:
Ah; good call on refactoring it out to a separate function :)